src/helper.cc

```diff
@@ -848,20 +870,28 @@ helper::handleKilledServer(HelperServerBase *srv, bool &needsNewServers)
         debugs(84, DBG_CRITICAL, "WARNING: " << id_name << " #" << srv->index << " exited");
 
         if (childs.needNew() > 0) {
```

Comment:
The other/new handleLackOfServers() code users do not check this _or similar_ condition. Should not they check something, especially since both of them say "the loop might start enough helpers"?! We should not log "too few .... processes are running" error if _enough_ of them are actually running.

Reply:
However, two other handleLackOfServers() calls check the `sawFailures` contition. If `sawFailures` is false, then either needNew() returned zero or there was a failure inside the loop (implying that some of those process have not been started). In other words, there is an implicit condition there meaning the same as this 'childs.needNew() > 0' check.

Reply:
> there is an implicit condition there meaning the same as this 'childs.needNew() > 0' check.

I think you are correct, but

* this conclusion contradicts our own code that adds code to delay handleLackOfServers() code because, and I quote, "the loop might start **enough** helpers" and
* this fragile implicit dependency may change without developers realizing that they broke something. We may start calling handleLackOfServers() from other contexts, for example.

To address these problems, I think we should make two changes:

* Change our comments from "; the loop might start enough helpers" to "until the final number of helpers is known". Alternatively, count successfully launched helpers (instead of remembering that we saw a failure) and use that counter to detect the "lack of servers" condition after the loop. That alternative is probably a bit more robust because it has a lower chance of missing a helper startup failure as the loop code gets altered. It also does not need a comment :-).

* (Double) check needNew() in handleLackOfServers() to avoid introducing a hidden dependency.

Reply:
>  count successfully launched helpers

Ok, let's do that.